File: 15_Semantic_Rules/ets_cases/15.6_Compatibility_of_Call_Arguments/cross_lang_verify/SEM_15_06_CallArgs.java
/**
 * Java cross-language verification for 15.6 Compatibility of Call Arguments
 *
 * Key findings: Very similar across all three.
 * - Parameter type must be compatible
 * - Subtype parameter passing OK
 * - Type mismatch → compile error
 */
class SEM_15_06_CallArgs {
    static int pass = 0, fail = 0;
    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    static class Animal {}
    static class Dog extends Animal {}

    static void acceptAnimal(Animal a) {}

    public static void main(String[] args) {
        // Subtype parameter passing
        acceptAnimal(new Dog());
        check(true, "subtype parameter passing (Java: same)");

        // Basic parameter passing
        acceptAnimal(new Animal());
        check(true, "compatible parameter type (Java: same)");

        // Type mismatch: compile error
        // acceptAnimal("string");  // COMPILE ERROR in Java
        check(true, "type mismatch rejection (Java: same)");

        // Parameter count mismatch: compile error
        // acceptAnimal();  // COMPILE ERROR
        check(true, "parameter count mismatch (Java: same)");

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) System.exit(1);
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.6_Compatibility_of_Call_Arguments/cross_lang_verify/SEM_15_06_CallArgs.swift
/**
 * Swift cross-language verification for 15.6 Compatibility of Call Arguments
 *
 * Key findings: Same as ArkTS/Java.
 */
import Foundation

var pass = 0, fail = 0
func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

class Animal {}
class Dog: Animal {}

func acceptAnimal(_ a: Animal) {}

// Subtype parameter passing
acceptAnimal(Dog())
check(true, "subtype parameter passing (Swift: same)")

// Basic parameter passing
acceptAnimal(Animal())
check(true, "compatible parameter type (Swift: same)")

// Type mismatch: compile error
// acceptAnimal("string")  // COMPILE ERROR
check(true, "type mismatch rejection (Swift: same)")

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")
